### ISSUE TITLE
New bankswitching type based on 3E (3EX)

### ISSUE THREAD

**Issue by thrust26** [stella-emu/stella#619]
3E is technical able to support 256K RAM. However the current implementation limits this to 32K (the RAM of the Krokodile Card). To allow using the full 256K RAM a new type should be defined.

Open questions:
- Name (e.g. 3EX)
- auto detection of new type and required RAM 

@sa666666, @DirtyHairy ,@andrew-davie

**Comment by DirtyHairy** [comment]
Concerning detection: I suggest we write the required number of RAM banks to both 0x0ffa and 0x0ffb. Together with "normal" 3E autodetection, this should be enough to detect the new scheme.

**Comment by thrust26** [comment]
Hm, how do we know that the values in 0x0ffa/b are valid values? 3E has values there too.

I would rather suggest a short string e.g. "3EX" followed by the number of RAM banks - 1.

**Comment by DirtyHairy** [comment]
My proposition is to store *the same* byte at both 0x0ffa and 0x0ffb. I would assume this to be specific enough to tell 3EX from regular 3E cartridges but, of course, a signature string will be more specific (more so if we restrict it to the first 4k).

**Comment by thrust26** [comment]
@andrew-davie
I have added 3EX to the refactor/cart branch, for auto detection the ROM needs the string "3EX" at least twice. 

TODOs: 
- debugger enhancements
- RAM size limitation

**Comment by DirtyHairy** [comment]
I suggest we read the RAM size from 0x0ffa.

**Comment by thrust26** [comment]
Agreed, the last but 6th byte (0x0ffa) of the ROM will contain the RAM bank count - 1.

**Comment by andrew-davie** [comment]
> Agreed, the last but 6th byte (0x0ffa) of the ROM will contain the RAM bank count - 1.

I have my doubts about the need to allow specification of (say) 57K of RAM. There should be some "chunk size" and reserve some bits in that byte for other things, maybe. I can't think of a use-case, but perhaps one bit to indicate how to handle RAM bank wrapping (illegal bank number).... either mask and wrap, or fault/reset for example?

You could 'chunk' it at 16K, use 4 bits to indicate RAM size, and the other 4 bits for... ?

One bit could even be reserved for "lda (zp),y" cannot cross pages, and thus allow the scheme to be able to run current 3E binaries with a patch to 0x0ffa only required.  That is, the bad apple would have 0x0ffa set to 32K RAM, wrapping, "lda (zp),y"... and it would work just like a KK.

Just throwing some ideas. Would be nice to have 3EX as a superset, in other words, that would also be back-compatible to 3E.  Not sure if anyone's going to like that idea.

**Comment by thrust26** [comment]
No clue about the RAM bank size. I figure it would be more something like 1K, 2K, 4K, 8K etc. which is supported by hardware. And as much as I understand hardware, the addresses would wrap around. I don't think there is hardware which behaves differently here.

Anyway, as of now this is preliminary only. There exists no hardware which fully supports that yet. When as specific request arrives, we can tailor it to the hardware needs (without giving up flexibility of course). I figured that you might want to use this for one of your games, no?

**Comment by andrew-davie** [comment]
> I figured that you might want to use this for one of your games, no?

As soon as something's available I will move Chess over to the new scheme. The extra RAM, though not needed, might come in handy some day. And besides it would be useful to kick-start support of the scheme in emulators and hardware.

**Comment by thrust26** [comment]
> As soon as something's available

You mean hardware? Or an official Stella release?

**Comment by andrew-davie** [comment]
> > As soon as something's available...
> 
> You mean hardware? Or an official Stella release?

Stella. Other emulators/hardware will follow.

**Comment by thrust26** [comment]
It will most likely be part of the next major release. Which shouldn't take as long as the last one.

**Comment by sa666666** [comment]
> It will most likely be part of the next major release. Which shouldn't take as long as the last one.

Definitely not.  The delay in the last release (over a year) is the longest gap between releases in Stella's 24 year history.  I don't want to duplicate that.  It was getting to the point where I forgot what changes happened earlier in the release cycle.

**Comment by sa666666** [comment]
Will this now be part of 6.2, or are we bumping it?  If so, we need to update the changelog and remove reference to it.  I'd rather not release it until it's done, so we don't run into the issue of ROMs being created while the definition is in flux.

**Comment by sa666666** [comment]
Well, it needs to have minimum functionality before we release it.  If you confident that it does, then we can include it.  Or finish up what needs to be done before the 6.2 release.  Your choice.

**Comment by thrust26** [comment]
It should work, but it is missing a debugger widget. I will work on that. Should be easy.

**Comment by thrust26** [comment]
Closed for now. If someone ever want to use it, we can further detail it.
